File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-7.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree2(self, preorder: List[int], inorder: List[int], limit) -> Optional[TreeNode]:
        if self.preIndx >= len(preorder):
            return None
        
        if inorder[self.inIndx] == limit:
            self.inIndx+=1
            return None
        
        root = TreeNode(preorder[self.preIndx])
        self.preIndx+=1
        root.left = self.buildTree2(preorder, inorder, root.val)
        root.right = self.buildTree2(preorder, inorder, limit)

        return root


    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        self.preIndx=0
        self.inIndx=0
        return self.buildTree2(preorder, inorder, sys.maxsize)

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-7.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int max=Integer.MIN_VALUE;
    public int maxMan(TreeNode root) {
        if(root == null) {
            return 0;
        }
        

        int maxLeft = maxMan(root.left);
        int maxRight = maxMan(root.right);
System.out.println("ARKA: " + root.val + " : " + maxLeft + " : " + maxRight);
System.out.println(max);
        max = Math.max(max, root.val);
System.out.println(max);
        max = Math.max(max, (maxLeft + maxRight + root.val));
System.out.println(max);
        if(root.left != null)
            max = Math.max(max, maxLeft + root.val);
System.out.println(max);
        if(root.right != null)
            max = Math.max(max, maxRight + root.val);
System.out.println(max);

        return Math.max(Math.max(root.val, root.val+maxLeft), root.val+maxRight);
    }
    public int maxPathSum(TreeNode root) {
        maxMan(root);
        return max;
    }
}
